packages/react/src/xcss-prop/index.ts

```diff
@@ -0,0 +1,143 @@
+ * your component.
+ */
+export type XCSSAllPseudos = CSSPseudos;
+
+/**
+ * ## xcss prop
+ *
+ * Declare styles your component takes with all other styles marked as violations
+ * by the TypeScript compiler. There are two primary use cases for xcss prop:
+ *
+ * - safe style overrides
+ * - inverting style declarations
+ *
+ * Interverting style declarations is interesting for platform teams as
+ * it means products only pay for styles they use as they're now the ones who declare
+ * the styles!
+ *
+ * The {@link XCSSProp} type has generics which must be defined — of which should be what you
+ * explicitly want to maintain as API. Use {@link XCSSAllProperties} and {@link XCSSAllPseudos}
+ * to enable all properties and pseudos.
+ *
+ * @example
+ * ```
+ * interface MyComponentProps {
+ *   // Color is accepted, all other properties / pseudos are considered violations.
+ *   xcss?: XCSSProp<'color', never>;
+ *
+ *   // Only backgrond color and hover pseudo is accepted.
+ *   xcss?: XCSSProp<'backgroundColor', '&:hover'>;
+ *
+ *   // All properties are accepted, all pseudos are considered violations.
+ *   xcss?: XCSSProp<XCSSAllProperties, never>;
+ *
+ *   // All properties are accepted, only the hover pseudo is accepted.
+ *   xcss?: XCSSProp<XCSSAllProperties, '&:hover'>;
+ * }
+ *
+ * function MyComponent({ xcss }: MyComponentProps) {
+ *   return <div css={{ color: 'var(--ds-text-danger)' }} className={xcss} />
+ * }
+ * ```
+ *
+ * The xcss prop works with static inline objects and the [cssMap](https://compiledcssinjs.com/docs/api-cssmap) API.
+ *
+ * @example
+ * ```
+ * // Declared as an inline object
+ * <Component xcss={{ color: 'var(--ds-text)' }} />
+ *
+ * // Declared with the cssMap API
+ * const styles = cssMap({ text: { color: 'var(--ds-text)' } });
+ * <Component xcss={styles.text} />
+ * ```
+ *
+ * To concatenate and conditonally apply styles use the {@link cssMap} {@link cx} functions.
+ */
+export type XCSSProp<
+  TAllowedProperties extends keyof CSSProperties,
+  TAllowedPseudos extends CSSPseudos
+> =
+  | (XCSSItem<TAllowedProperties> & XCSSPseudos<TAllowedProperties, TAllowedPseudos> & XCSSAtRules)
+  | false
+  | null
+  | undefined;
+
+/**
+ * ## cx
+ *
+ * Use in conjunction with the {@link XCSSProp} to concatenate and conditionally apply
+ * declared styles. Can only be used with the `cssMap()` and {@link XCSSProp} APIs.
+ *
+ * @example
+ * ```
+ * const styles = cssMap({
+ *  text: { color: 'var(--ds-text)' },
+ *  primary: { color: 'var(--ds-text-brand)' },
+ * });
+ *
+ * <Component xcss={cx(isPrimary && styles.text, !isPrimary && styles.primary)} />
+ * ```
+ */
+export const cx = <TStyles extends [...XCSSProp<any, any>[]]>(
+  ...styles: TStyles
+): TStyles[number] & string => {
+  // Types won't match here as type-time are always objects.
+  // At runtime however they will be an array of strings.
+  return ac(styles as unknown as string[]) as unknown as TStyles[number] & string;
```

Comment:
Without diving in too much as this is all my first time seeing the _rest_ of the Compiled code, this does feel wrong of course, but I can't seem to fix it in 30s in a codespace, so I'll just believe the comment.

I will note that `ac` doesn't support `null`, eg. `classNames: (AtomicGroups | string | undefined | false)[]` — will it be handled fine, eg. is there a `typeof === 'object'` that might trigger on a null?

Reply:
😅 

So `ac` doesn't return a string, it returns a class that has a `toString()` method. To make sure we can just pass it to areas that accept strings (...like `className`) we intersect it with the `string` literal.

@liamqma is `ac` safe taking `null`? To Kylors point around the typeof check.

- I'll look into adding `null` to `ac` type.

Reply:
ac should handle `null` as it's doing truthy check. 

> I'll look into adding null to ac type.
Thank you. Can you add it to ax as well. It should be the same.

Reply:
Pushed up improvements.